### PR TITLE
Add user info editor page

### PR DESCRIPTION
**Pull request by Illusion4**
https://github.com/BinaryStudioAcademy/bsa-2023-leetwars/assets/88484289/c4ec378d-ca6a-4fcd-bdf5-ac7391c9b5b0

### CODE REVIEW

**Comment by Illusion4** [comment]
[Edit user info component](https://trello.com/c/7sNYIAFT/134-edit-user-info-component)

**Review by bodichka11 — APPROVED** [review]

File: backend/LeetWars.Core/LeetWars.Core.BLL/Services/UserService.cs [hunk 1]
```diff
@@ -170,6 +175,44 @@ public async Task<List<UserDto>> GetLeaderBoardAsync(PageSettingsDto? page)
...
+        currentUser!.UserName = userInfoDto.Username!;
+        
+        _context.Users.Update(currentUser);
+        await _context.SaveChangesAsync();
+        return _mapper.Map<UserDto>(currentUser);
+    }
+
+    public async Task<UserAvatarDto> UpdateUserAvatar(IFormFile image)
+    {
+        if(image is null)
```
Comment (tatianahutii) [inline backend/LeetWars.Core/LeetWars.Core.BLL/Services/UserService.cs]:
```suggestion
        if (image is null)

```

File: backend/LeetWars.Core/LeetWars.Core.BLL/Services/UserService.cs [hunk 2]
```diff
@@ -170,6 +175,44 @@ public async Task<List<UserDto>> GetLeaderBoardAsync(PageSettingsDto? page)
...
+    public async Task<UserDto> UpdateUserInfo(UpdateUserInfoDto userInfoDto)
+    {
+        if (userInfoDto is null)
+        {
+            throw new ArgumentNullException(nameof(userInfoDto));
+        }
+        
+        var currentUser = await _context.Users.FirstOrDefaultAsync(x => x.Uid == _userGetter.CurrentUserId);
+        
+        currentUser!.Email = userInfoDto.Email!;
```
Comment (tatianahutii) [inline backend/LeetWars.Core/LeetWars.Core.BLL/Services/UserService.cs]: please avoid using !

File: backend/LeetWars.Core/Dockerfile [hunk 3]
```diff
@@ -1,5 +1,13 @@
 FROM mcr.microsoft.com/dotnet/aspnet:6.0-focal AS base
 ARG PROJECT_PORT
+ARG BLOB_URL
+ARG BLOB_CONTAINER_NAME
+ARG BLOB_ACCESS
+
+ENV BLOB_URL=$BLOB_URL
+ENV BLOB_CONTAINER_NAME=$BLOB_CONTAINER_NAME
+ENV BLOB_ACCESS=$BLOB_ACCESS
+
```
Comment (tatianahutii) [inline backend/LeetWars.Core/Dockerfile]: do we really need this? we can store all secret data in env variables, not secret in appsettings

File: frontend/src/app/modules/user/user-info-editor/user-info-editor.component.ts [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+                return user;
+            }),
+        );
+    }
+
+    private loadData() {
+        this.userService.getCurrentUser().subscribe((user) => {
+            this.user = user;
+            this.initializeForm(user);
+            this.avatarPreview = this.user?.imagePath ?? '/assets/images/default-profile-img.jpg';
```
Comment (tatianahutii) [inline frontend/src/app/modules/user/user-info-editor/user-info-editor.component.ts]: magic string

File: frontend/src/app/modules/user/user-info-editor/user-info-editor.component.ts [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+            email: user!.email,
+            username: user!.userName,
+        });
+    }
+
+    private updateAvatarPreview(target: HTMLInputElement) {
+        if (!target.files || !target.files[0]) {
+            return;
+        }
+        const file = target.files[0];
```
Comment (tatianahutii) [inline frontend/src/app/modules/user/user-info-editor/user-info-editor.component.ts]:
```suggestion
        const file =  target.files && target.files[0];
         if (!file ) {
            return;
        }
```

File: frontend/src/app/shared/components/header/header.component.ts [hunk 6]
```diff
@@ -25,9 +28,10 @@ export class HeaderComponent {
         private headerService: HeaderService,
         private toastrService: ToastrNotificationsService,
     ) {
-        this.authService.getUser().subscribe((user) => {
-            this.user = user;
-        });
+        this.authService.currentUser$.pipe(takeUntil(this.destroy$))
```
Comment (tatianahutii) [inline frontend/src/app/shared/components/header/header.component.ts]: you can inherit BaseComponent

File: frontend/src/app/shared/models/user/edit-user-info.ts [hunk 7]
```diff
@@ -0,0 +1,4 @@
+export interface EditUserInfo {
```
Comment (tatianahutii) [inline frontend/src/app/shared/models/user/edit-user-info.ts]:
```suggestion
export interface IEditUserInfo {

```

File: frontend/src/app/shared/models/user/user-avatar.ts [hunk 8]
```diff
@@ -0,0 +1,3 @@
+export interface UserAvatar {
```
Comment (tatianahutii) [inline frontend/src/app/shared/models/user/user-avatar.ts]:
```suggestion
export interface IUserAvatar {

```

**Review by tatianahutii — CHANGES_REQUESTED** [review]

File: backend/LeetWars.Core/LeetWars.Core.BLL/Services/UserService.cs [hunk 9]
```diff
@@ -170,6 +175,46 @@ public async Task<List<UserDto>> GetLeaderBoardAsync(PageSettingsDto? page)
...
+        if (userInfoDto is null)
+        {
+            throw new ArgumentNullException(nameof(userInfoDto));
+        }
+
+        var currentUser = await _context.Users.FirstOrDefaultAsync(x => x.Uid == _userGetter.CurrentUserId)
+                            ?? throw new InvalidOperationException($"A user with id {_userGetter.CurrentUserId} is not found.");
+        
+        currentUser.Email = userInfoDto.Email;
+        currentUser.UserName = userInfoDto.Username;
```
Comment (tatianahutii) [inline backend/LeetWars.Core/LeetWars.Core.BLL/Services/UserService.cs]: auto mapper?

File: backend/LeetWars.Core/LeetWars.Core.BLL/Services/UserService.cs [hunk 10]
```diff
@@ -170,6 +175,46 @@ public async Task<List<UserDto>> GetLeaderBoardAsync(PageSettingsDto? page)
...
+    }
+
+    public async Task<UserAvatarDto> UpdateUserAvatar(IFormFile image)
+    {
+        if (image is null)
+        {
+            throw new ArgumentNullException(nameof(image));
+        }
+        
+        var currentUser = await _context.Users.FirstOrDefaultAsync(x => x.Uid == _userGetter.CurrentUserId)
```
Comment (tatianahutii) [inline backend/LeetWars.Core/LeetWars.Core.BLL/Services/UserService.cs:222]: Don't we have our custom exceptions for this case?

File: backend/LeetWars.Core/LeetWars.Core.Common/DTO/User/UserFullDto.cs [hunk 11]
```diff
@@ -20,7 +20,7 @@ public class UserFullDto : Entity<long>
         public string LastName { get; set; } = string.Empty;
         public string UserName { get; set; } = string.Empty;
         public string Email { get; set; } = string.Empty;
-        public string ImagePath { get; set; } = string.Empty;
+        public string? ImagePath { get; set; }
```
Comment (tatianahutii) [inline backend/LeetWars.Core/LeetWars.Core.Common/DTO/User/UserFullDto.cs:23]: is it duplication? we have ImagePath in IDtoWithImage as well

File: backend/LeetWars.Core/LeetWars.Core.Common/DTO/User/UserDto.cs [hunk 12]
```diff
@@ -3,7 +3,7 @@
 
 namespace LeetWars.Core.Common.DTO.User
 {
-    public class UserDto : Entity<long>
+    public class UserDto : Entity<long>, IDtoWithImage
```
Comment (tatianahutii) [inline backend/LeetWars.Core/LeetWars.Core.Common/DTO/User/UserDto.cs:6]: the same here

File: frontend/src/app/modules/user/user-info-editor/user-info-editor.component.sass [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    width: 186px
+    background-color: $green
+    border-radius: 10px
+    font-weight: 800
+
+.form-wrapper
+    max-width: 970px
+    gap: 39px
+
+.editUserInfoForm
```
Comment (tatianahutii) [inline frontend/src/app/modules/user/user-info-editor/user-info-editor.component.sass]: use the same style for class names

**Review by tatianahutii — CHANGES_REQUESTED** [review]

**Review by tatianahutii — APPROVED** [review]

**Review by apofis96 — APPROVED** [review]

**Pull request merged**